easybuild/easyblocks/f/flang.py

```diff
@@ -0,0 +1,186 @@
+    """Support for bootstrapping Flang."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize custom class variables for Flang."""
+
+        super(EB_Flang, self).__init__(*args, **kwargs)
+
+    def disable_sanitizer_tests(self):
+        # Not relevant for flang
+        pass
+
+    def extract_step(self):
+        """
+        Prepare a combined LLVM source tree.  The layout is:
+        llvm/             Unpack flang-llvm-*.tar.gz here
+          projects/
+            openmp/       Unpack openmp-*.tar.xz here
+          tools/
+            clang/        Unpack flang-flang-driver*.tar.gz here
+        """
+
+        # Extract everything into separate directories.
+        super(EB_Clang, self).extract_step()
+
+        # Find the full path to the directory that was unpacked from flang-llvm-*.tar.gz
+        for tmp in self.src:
+            if tmp['name'].startswith("flang-llvm-"):
+                self.llvm_src_dir = tmp['finalpath']
+                break
+
+        if self.llvm_src_dir is None:
+            raise EasyBuildError(
+                "Could not determine LLVM source root (LLVM source was not unpacked?)")
+        src_dirs = {}
+
+        def find_source_dir(globpatterns, targetdir):
+            """Search for directory with globpattern and rename it to targetdir"""
+            if not isinstance(globpatterns, list):
+                globpatterns = [globpatterns]
+
+            glob_src_dirs = [glob_dir for globpattern in globpatterns for glob_dir in
+                             glob.glob(globpattern)]
+            if len(glob_src_dirs) != 1:
+                raise EasyBuildError(
+                    "Failed to find exactly one source directory for pattern %s: %s",
+                    globpatterns,
+                    glob_src_dirs)
+            src_dirs[glob_src_dirs[0]] = targetdir
+
+        if LooseVersion(self.version) >= LooseVersion('3.8'):
+            find_source_dir('openmp-*',
+                            os.path.join(self.llvm_src_dir, 'projects', 'openmp'))
+
+        find_source_dir('flang-driver-*',
+                        os.path.join(self.llvm_src_dir, 'tools', 'clang'))
+
+        # Place the flang code in a separate directory for the build step for after
+        # we've built llvm
+        self.flang_source_dir = os.path.join(self.llvm_src_dir, '..', 'flang')
+        find_source_dir('flang-flang_*', self.flang_source_dir)
+
+        for src in self.src:
+            for (dirname, new_path) in src_dirs.items():
+                if src['name'].startswith(dirname):
+                    old_path = os.path.join(src['finalpath'], dirname)
+                    try:
+                        shutil.move(old_path, new_path)
+                    except IOError as err:
+                        raise EasyBuildError("Failed to move %s to %s: %s", old_path,
+                                             new_path, err)
+                    src['finalpath'] = new_path
+                    break
```

Comment:
Shouldn't we somehow checked that all entries in `self.src` got moved? There's no guarantee of that now...